processor/src/main/java/org/mapstruct/ap/internal/util/SpecificCompilerWorkarounds.java

```diff
@@ -115,4 +135,161 @@ public static TypeElement replaceTypeElementIfNecessary(Elements elementUtils, T
         }
         return element;
     }
+
+    public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                        Elements elementUtils) {
+        try {
+            // quick and dirty, until implementation below is more sound
+            return typeUtils.asMemberOf( containing, element );
+        }
+        catch ( IllegalArgumentException e ) {
+            TypeMirror eclipseWorkaround =
+                EclipseEnvironment.asMemberOf( containing, element, typeUtils, elementUtils );
+
+            if ( null != eclipseWorkaround ) {
+                return eclipseWorkaround;
+            }
+
+            throw e;
+        }
+    }
+
+    private static final class EclipseEnvironment {
+        public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                            Elements elementUtils) {
+            if ( typeUtils instanceof TypesImpl && containing instanceof DeclaredTypeImpl
+                && element instanceof ElementImpl ) {
+                try {
+                    ElementImpl elementImpl = (ElementImpl) element;
+
+                    Field field = TypesImpl.class.getDeclaredField( "_env" );
+                    field.setAccessible( true );
+                    BaseProcessingEnvImpl env = (BaseProcessingEnvImpl) field.get( typeUtils );
+
+                    // TODO can some of the reflection here be avoided?
+                    Method bindingGetter = TypeMirrorImpl.class.getDeclaredMethod( "binding" );
+                    bindingGetter.setAccessible( true );
+                    ReferenceBinding referenceBinding = (ReferenceBinding) bindingGetter.invoke( containing );
+
+                    MethodBinding methodBinding = (MethodBinding) elementImpl._binding;
+                    // matches in super-classes have priority
+                    MethodBinding inSuperclassHiearchy =
+                        findInSuperclassHierarchy( methodBinding, env, referenceBinding );
+                    if ( inSuperclassHiearchy != null ) {
+                        return env.getFactory().newTypeMirror( inSuperclassHiearchy );
+                    }
+
+                    // if nothing was found, traverse the interfaces and collect all candidate methods that match
+                    List<MethodBinding> candidatesFromInterfaces = new ArrayList<MethodBinding>();
+                    collectFromInterfaces(
+                        methodBinding,
+                        referenceBinding,
+                        new HashSet<ReferenceBinding>(),
+                        candidatesFromInterfaces );
+
+                    java.util.Collections.sort( candidatesFromInterfaces, new Comparator<MethodBinding>() {
```

Comment:
In case you find the same method provided in two implemented interfaces. That's what I mumbled about in the Gitter-chat. I still need to create a test case to check if that implementation here even works.